test/integration/consul-container/test/ratelimit/ratelimit_test.go

```diff
@@ -0,0 +1,261 @@
+
+			client, err := cluster.GetClient(nil, true)
+			require.NoError(t, err)
+
+			// perform actions and validate returned errors to client
+			for _, op := range tc.operations {
+				err = op.action.function(client)
+				if len(op.expectedErrorMsg) > 0 {
+					require.Error(t, err)
+					require.Equal(t, op.expectedErrorMsg, err.Error())
+				} else {
+					require.NoError(t, err)
+				}
+			}
+
+			// validate logs and metrics
+			// doing this in a separate loop so we can perform actions, allow metrics
+			// and logs to collect and then assert on each.
+			for _, op := range tc.operations {
+				// validate metrics
+				metricsInfo, err := client.Agent().Metrics()
+				// TODO(NET-1978): currently returns NaN error
+				//			require.NoError(t, err)
+				if metricsInfo != nil && err == nil {
+					if op.expectMetric {
+						checkForMetric(t, metricsInfo, op.action.rateLimitOperation, op.action.rateLimitType)
+					}
+				}
+
+				// validate logs
+				// putting this last as there are cases where logs
+				// were not present in consumer when assertion was made.
+				if op.expectExceededLog {
+					checkLogsForMessage(t, logConsumer.Msgs,
+						fmt.Sprintf("[TRACE] agent.server.rpc-rate-limit: RPC exceeded allowed rate limit: rpc=%s", op.action.rateLimitOperation),
+						op.action.rateLimitOperation, "exceeded")
+				}
+				if op.expectBlockedLog {
+					checkLogsForMessage(t, logConsumer.Msgs,
+						fmt.Sprintf("[WARN]  agent.server.rpc-rate-limit: RPC blocked due to exceeded allowed rate limit: rpc=%s", op.action.rateLimitOperation),
+						op.action.rateLimitOperation, "blocked")
+				}
+			}
+		})
+	}
+}
+
+func checkForMetric(t *testing.T, metricsInfo *api.MetricsInfo, operationName string, expectedLimitType string) {
+	for _, counter := range metricsInfo.Counters {
+		if counter.Name == "consul.rate.limit" {
+			operation, ok := counter.Labels["op"]
+			require.True(t, ok)
+
+			limitType, ok := counter.Labels["limit_type"]
+			require.True(t, ok)
+
+			mode, ok := counter.Labels["mode"]
+			require.True(t, ok)
+
+			if operation == operationName {
+				require.Equal(t, 2, counter.Count)
+				require.Equal(t, expectedLimitType, limitType)
+				require.Equal(t, "disabled", mode)
+			}
+		}
+	}
+
+}
+
+func checkLogsForMessage(t *testing.T, logs []string, msg string, operationName string, logType string) {
+	found := false
+	for _, log := range logs {
+		if strings.Contains(log, msg) {
+			found = true
+		}
+	}
+	require.True(t, found, fmt.Sprintf("%s log not found for: %s", logType, operationName))
+}
+
+func terminate(t *testing.T, cluster *libcluster.Cluster) {
+	err := cluster.Terminate()
+	require.NoError(t, err)
+}
+
+type TestLogConsumer struct {
+	Msgs []string
+}
+
+func (g *TestLogConsumer) Accept(l testcontainers.Log) {
+	g.Msgs = append(g.Msgs, string(l.Content))
+}
+
+// createCluster
+func createCluster(t *testing.T, cmd string, logConsumer *TestLogConsumer, logLevel string) *libcluster.Cluster {
```

Comment:
We have a `createCluster` func in `test/integration/consul-container/test/basic/connect_service_test.go`, I'm wondering if we can unify both and extract it from the test file?

Reply:
I think I am going to backlog a ticket so we can get Xinyi or Poonam to work on this.  It would be helpful to get them oriented to these integration tests.